<a href="https://colab.research.google.com/github/evgeniytr1509/HW_DS_7/blob/main/HW_DS7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [27]:
!pip install scikit-surprise

In [29]:
from surprise import SVD, SVDpp, NMF
from surprise import Dataset
from surprise import accuracy
from surprise.model_selection import cross_validate
from scipy.io import loadmat
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
data = Dataset.load_builtin('ml-100k')

In [30]:
data

Алгоритм SVD

In [31]:
# Создайте объект модели SVD
model = SVD()

In [48]:
#  кросс-валидациz
resultSVD = cross_validate(model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9206  0.9181  0.9181  0.9269  0.9189  0.9205  0.0033  
MAE (testset)     0.7210  0.7211  0.7191  0.7266  0.7205  0.7217  0.0026  
Fit time          26.74   26.90   27.93   27.68   27.85   27.42   0.50    
Test time         4.60    4.27    5.24    4.47    5.06    4.73    0.37    


In [33]:
# Выведите результаты
for measure in ['test_rmse', 'test_mae']:
    print(f'{measure}: {resultSVD[measure].mean()}')

test_rmse: 0.9354176760616545
test_mae: 0.737553476664323


In [34]:
algoSVDpp = SVDpp()

Алгоритм SVD++

In [47]:
# объект модели SVD++
model = SVDpp()

#  кросс-валидация
results = cross_validate(algoSVDpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)


for measure in ['test_rmse', 'test_mae']:
  print(f'{measure}: {results[measure].mean()}')

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9112  0.9163  0.9199  0.9248  0.9187  0.9182  0.0045  
MAE (testset)     0.7144  0.7188  0.7210  0.7245  0.7213  0.7200  0.0033  
Fit time          28.17   26.81   28.09   26.77   27.22   27.41   0.61    
Test time         4.94    4.18    4.51    4.42    4.36    4.48    0.25    
test_rmse: 0.9181690172663183
test_mae: 0.7200071561368272


Алгоритм NMF

In [41]:
from surprise import NMF

# модель NMF
model = NMF()

# кросс-валидацию
results = cross_validate(model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

for measure in ['test_rmse', 'test_mae']:
   print(f'{measure}: {results[measure].mean()}')

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9622  0.9687  0.9664  0.9672  0.9518  0.9633  0.0061  
MAE (testset)     0.7579  0.7617  0.7591  0.7597  0.7494  0.7576  0.0043  
Fit time          2.65    2.20    2.12    2.09    2.19    2.25    0.20    
Test time         0.19    0.25    0.10    0.24    0.11    0.18    0.06    
test_rmse: 0.963253844712785
test_mae: 0.7575588355555023


In [50]:
from surprise import Dataset, Reader
from surprise import SVD, SVDpp, NMF
from surprise.model_selection import cross_validate


# Создайте модели
models = [SVD(), SVDpp(), NMF()]

# кросс-валидацию для каждой модели
results = []
for model in models:
    cv_results = cross_validate(model, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
    results.append(cv_results)

# результаты для каждой модели
for i, model in enumerate(models):
    print(f"Модель: {model.__class__.__name__}")
    for measure in ['test_rmse', 'test_mae']:
        print(f'{measure}: {results[i][measure].mean()}')
    print()


Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9348  0.9375  0.9373  0.9391  0.9384  0.9374  0.0015  
MAE (testset)     0.7383  0.7401  0.7386  0.7374  0.7423  0.7393  0.0017  
Fit time          3.02    1.40    1.38    1.36    1.36    1.70    0.66    
Test time         0.35    0.13    0.26    0.14    0.15    0.21    0.09    
Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9182  0.9195  0.9233  0.9198  0.9298  0.9221  0.0042  
MAE (testset)     0.7205  0.7221  0.7248  0.7219  0.7296  0.7238  0.0032  
Fit time          26.67   26.65   26.79   26.55   27.00   26.73   0.15    
Test time         5.40    4.12    5.17    4.06    4.99    4.75    0.55    
Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (tests

Алгоритм SVD++ является наилучшим выбором с точки зрения точности предсказаний (Среднее RMSE: 0.9221, - наименьшее) но заметно медленнее обучается и тестируется (среднее Fit time 26,73 Test time 4.75).

Алгоритм SVD также предоставляет хорошие результаты и более быструю скорость обучения и тестирования по сравнению с SVD++.

Алгоритм NMF имеет более высокие значения RMSE и MAE по сравнению с SVD и SVD++, что может указывать на менее точные предсказания. Однако, время обучения и тестирования NMF сравнительно небольшое.

Наиболее оптимальным вариантом предлагается выбрать алгоритм SVD++ с точки зрения точности предсказаний. Выбор наиболее оптимального алгоритма зависит от требований технического задания, а именно какие параметры модели имеют более высокий приоритет.

**Задание 2**

In [81]:
url = 'https://drive.google.com/uc?id=189qpCpuWaVPC1V_W0J_Wld-z8LKg_f0A'
data = pd.read_csv(url, sep='\t', header=None, encoding='ISO-8859-1')
data

0
0                                 1 Toy Story (1995)
1                                 2 GoldenEye (1995)
2                                3 Four Rooms (1995)
3                                4 Get Shorty (1995)
4                                   5 Copycat (1995)
...                                              ...
1677                          1678 Mat' i syn (1997)
1678                           1679 B. Monkey (1998)
1679                       1680 Sliding Doors (1998)
1680                        1681 You So Crazy (1994)
1681  1682 Scream of Stone (Schrei aus Stein) (1991)

[1682 rows x 1 columns]

In [86]:
def loadMovieList(movie_list):
    """
    Reads the fixed movie list in movie_ids.txt and returns a list of movie names.
    Returns
    -------
    movieNames : list
        A list of strings, representing all movie names.
    """
    # Read the fixed movieulary list
    with open(movie_list,  encoding='ISO-8859-1') as fid:
        movies = fid.readlines()

    movieNames = []
    for movie in movies:
        parts = movie.split()
        movieNames.append(' '.join(parts[1:]).strip())
    return movieNames

In [ ]:
names = loadMovieList('movie_ids.txt')

In [87]:
# Load data
data = loadmat('movies.mat')
Y, R = data['Y'], data['R']

# Y is a 1682x943 matrix, containing ratings (1-5) of
# 1682 movies on 943 users

# R is a 1682x943 matrix, where R(i,j) = 1
# if and only if user j gave a rating to movie i

# From the matrix, we can compute statistics like average rating.
print('Average rating for movie 1601 (', names[1600], '): %f / 5' %
      np.mean(Y[1600, R[0, :]]))

# We can "visualize" the ratings matrix by plotting it with imshow
plt.figure(figsize=(8, 8))
plt.imshow(Y)
plt.ylabel('Movies')
plt.xlabel('Users')
plt.grid(False)
plt.show()

FileNotFoundError: ignored